### PR TITLE
Support table model insert redirection

### PR DESCRIPTION
**Pull request by HTHou**
## Description

This PR implements the insert redirection for table model.

## Test result

On 3c3d IoTDB cluster.  3 machines with 16C 8G.

```
data_region_replication_factor=2
```

iot-benchmark config
```
----------------------Main Configurations----------------------
########### Test Mode ###########
BENCHMARK_WORK_MODE=testWithDefaultPath
IoTDB_DIALECT_MODE=TABLE
########### Database Connection Information ###########
DOUBLE_WRITE=false
DBConfig=
  DB_SWITCH=IoTDB-200-SESSION_BY_TABLET
  HOST=[192.168.130.14]
########### Data Mode ###########
GROUP_NUMBER=1
IoTDB_TABLE_NUMBER=1
DEVICE_NUMBER=6000
REAL_INSERT_RATE=1.0
SENSOR_NUMBER=200
IS_SENSOR_TS_ALIGNMENT=true
IS_OUT_OF_ORDER=true
OUT_OF_ORDER_RATIO=0.5
########### Data Amount ###########
OPERATION_PROPORTION=1:0:0:0:0:0:0:0:0:0:0:0
CLIENT_NUMBER=20
LOOP=100
BATCH_SIZE_PER_WRITE=100
DEVICE_NUM_PER_WRITE=1
START_TIME=2022-01-01T00:00:00+08:00
POINT_STEP=5000
OP_MIN_INTERVAL=0
OP_MIN_INTERVAL_RANDOM=false
INSERT_DATATYPE_PROPORTION=1:1:1:1:1:1:0:0:0:0
ENCODINGS=RLE/TS_2DIFF/TS_2DIFF/GORILLA/GORILLA/DICTIONARY/PLAIN/PLAIN/TS_2DIFF/TS_2DIFF
COMPRESSOR=LZ4
########### Other Param ###########
IS_DELETE_DATA=true
CREATE_SCHEMA=true
BENCHMARK_CLUSTER=false
VECTOR=true
```
Before this PR
```
----------------------------------------------------------Result Matrix----------------------------------------------------------
Operation                okOperation              okPoint                  failOperation            failPoint                throughput(point/s)      
INGESTION                600000                   12060000000              0                        0                        8182821.01  
--------------------------------------------------------------------------Latency (ms) Matrix--------------------------------------------------------------------------
Operation                AVG         MIN         P10         P25         MEDIAN      P75         P90         P95         P99         P999        MAX         SLOWEST_THREAD
INGESTION                46.70       2.06        5.65        11.35       21.50       29.70       45.10       108.78      867.89      2224.37     3672.33     1467584.47  
```

After then
```
----------------------------------------------------------Result Matrix----------------------------------------------------------
Operation                okOperation              okPoint                  failOperation            failPoint                throughput(point/s)      
INGESTION                600000                   12060000000              0                        0                        9459863.84   
--------------------------------------------------------------------------Latency (ms) Matrix--------------------------------------------------------------------------
Operation                AVG         MIN         P10         P25         MEDIAN      P75         P90         P95         P99         P999        MAX         SLOWEST_THREAD
INGESTION                40.41       1.43        3.47        4.55        8.65        18.68       27.87       101.22      984.72      2501.42     6322.80     1268322.81 
```

### CODE REVIEW

File: iotdb-client/session/src/test/java/org/apache/iotdb/session/SessionCacheLeaderTest.java [hunk 1]
```diff
@@ -931,6 +1025,161 @@ public void testInsertTabletsWithSessionBroken() throws StatementExecutionExcept
...
+      }
+
+      if (tablet.rowSize == tablet.getMaxRowNumber()) {
+        try {
+          session.insertRelationalTablet(tablet);
+        } catch (IoTDBConnectionException e) {
+          assertEquals(
+              "the session connection = TEndPoint(ip:127.0.0.1, port:55560) is broken",
+              e.getMessage());
+        }
```
Comment (jt2594838) [inline iotdb-client/session/src/test/java/org/apache/iotdb/session/SessionCacheLeaderTest.java:1153]: What if the exception is not thrown?

**Review by jt2594838 — APPROVED** [review]

File: iotdb-client/isession/src/main/java/org/apache/iotdb/isession/IPooledSession.java [hunk 2]
```diff
@@ -28,6 +28,7 @@
 import org.apache.thrift.TException;
 import org.apache.tsfile.write.record.Tablet;
 
+/** NOTICE: IPooledSession is specific to the table model. */
 public interface IPooledSession extends AutoCloseable {
```
[the remark this answers is not in the document]
Comment (HTHou) [reply iotdb-client/isession/src/main/java/org/apache/iotdb/isession/IPooledSession.java:32]: The annotation @TableModel is in iotdb-server module. It is hard to use it in session and isession module, due to that iotdb-server depends on them. 
https://github.com/apache/iotdb/blob/374deefeb86793643f49a623fe9d6e05b4212bd7/iotdb-core/datanode/pom.xml#L72-L76

[the remark this answers is not in the document]
Comment (jt2594838) [reply iotdb-client/isession/src/main/java/org/apache/iotdb/isession/IPooledSession.java:32]: How about moving it to the common module? If session depends on it, too.

File: iotdb-client/session/src/main/java/org/apache/iotdb/session/Session.java [hunk 3]
```diff
@@ -1384,6 +1440,34 @@ private void handleRedirection(String deviceId, TEndPoint endpoint) {
...
+                  return constructSessionConnection(this, endpoint, zoneId);
+                } catch (IoTDBConnectionException ex) {
+                  exceptionReference.set(ex);
+                  return null;
+                }
+              });
+      if (connection == null) {
+        tableModelDeviceIdToEndpoint.remove(deviceId);
+      }
+    }
```
[the remark this answers is not in the document]
Comment (HTHou) [reply iotdb-client/session/src/main/java/org/apache/iotdb/session/Session.java:1468]: no idea either, just copied from old code. Let me remove it

File: iotdb-client/session/src/main/java/org/apache/iotdb/session/Session.java [hunk 4]
```diff
@@ -2697,6 +2790,111 @@ public void insertRelationalTablet(Tablet tablet)
...
+            });
+      }
+    }
+    insertRelationalTabletByGroup(relationalTabletGroup);
+  }
+
+  @SuppressWarnings({
+    "squid:S3776"
+  }) // ignore Cognitive Complexity of methods should not be too high
+  private <T> void insertRelationalTabletByGroup(
```
[the remark this answers is not in the document]
Comment (HTHou) [reply iotdb-client/session/src/main/java/org/apache/iotdb/session/Session.java:2832]: removed

File: iotdb-client/session/src/main/java/org/apache/iotdb/session/Session.java [hunk 5]
```diff
@@ -2697,6 +2790,111 @@ public void insertRelationalTablet(Tablet tablet)
...
+    for (CompletableFuture<Void> completableFuture : completableFutures) {
+      try {
+        completableFuture.join();
+      } catch (CompletionException completionException) {
+        Throwable cause = completionException.getCause();
+        logger.error("Meet error when async insert!", cause);
+        if (cause instanceof IoTDBConnectionException) {
+          throw (IoTDBConnectionException) cause;
+        } else {
+          errMsgBuilder.append(cause.getMessage());
```
[the remark this answers is not in the document]
Comment (HTHou) [reply iotdb-client/session/src/main/java/org/apache/iotdb/session/Session.java:2889]: Makes sense.

File: iotdb-client/session/src/test/java/org/apache/iotdb/session/SessionCacheLeaderTest.java (hunk 1 above)
Comment (HTHou) [reply iotdb-client/session/src/test/java/org/apache/iotdb/session/SessionCacheLeaderTest.java:1153]: Fixed

File: iotdb-client/session/src/main/java/org/apache/iotdb/session/pool/SessionPool.java [hunk 6]
```diff
@@ -703,7 +709,16 @@ private ISession getSession() throws IoTDBConnectionException {
...
+        if (sqlDialect.equals("tree")) {
+          session.open(
+              enableCompression, connectionTimeoutInMs, deviceIdToEndpoint, availableNodes);
+        } else {
+          session.open(
+              enableCompression,
+              connectionTimeoutInMs,
+              availableNodes,
+              tableModelDeviceIdToEndpoint);
+        }
```
[the remark this answers is not in the document]
Comment (HTHou) [reply iotdb-client/session/src/main/java/org/apache/iotdb/session/pool/SessionPool.java]: `sqlDialect.equals("tree")` removed

File: iotdb-client/isession/src/main/java/org/apache/iotdb/isession/IPooledSession.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (HTHou) [reply iotdb-client/isession/src/main/java/org/apache/iotdb/isession/IPooledSession.java:32]: No, session doesn't depends on node-commons module.

**Review by jt2594838 — APPROVED** [review]

**Pull request merged**